skglm/experimental/reweighted.py

```diff
@@ -0,0 +1,90 @@
+import numpy as np
+from skglm.estimators import GeneralizedLinearEstimator
+from skglm.penalties import WeightedL1, L0_5
+
+
+# def _log_sum_objective(coef):
+#     return np.log(np.abs(coef))
+
+
+# def _log_sum_derivative(coef):
+#     return 1. / (np.abs(coef))
+
+
+class IterativeReweightedL1(GeneralizedLinearEstimator):
+    r"""Reweighted L1-norm estimator.
+
+    This estimator iteratively solves a non-convex objective by iteratively solving
+    convex surrogates. They are obtained by iteratively re-weighting the \ell_1-norm
+    by a penalty applied to the coefficients.
+
+    Parameters
+    ----------
+    alpha : float, optional
+        Penalty strength.
+
+    datafit : instance of BaseDatafit, optional
+        Datafit. If None, `datafit` is initialized as a `Quadratic` datafit.
+        `datafit` is replaced by a JIT-compiled instance when calling fit.
+
+    solver : instance of BaseSolver, optional
+        Solver. If None, `solver` is initialized as an `AndersonCD` solver.
+
+    n_reweights : int, optional
+        Number of reweighting iterations.
+
+    Attributes
+    ----------
+    coef_ : array, shape (n_features,)
+        Parameter vector (w in the cost function formula)
+
+    loss_history_ : list
+        Objective history after every reweighting iteration
+
+    References
+    ----------
+    .. [1] Candès et al. (2007), Enhancing sparsity by reweighted l1 minimization
+           https://web.stanford.edu/~boyd/papers/pdf/rwl1.pdf
+    """
+
+    def __init__(self, alpha, datafit=None, penalty=L0_5(1.), solver=None,
+                 n_reweights=5):
+        super().__init__(
+            datafit=datafit, penalty=WeightedL1(alpha, np.ones(1)), solver=solver)
+        self.n_reweights = n_reweights
+        self.nncvx_penalty = penalty
```

Comment:
Do you have a better name? Any idea on how to handle both penalties. I find this naming messy.

Reply:
use penalty for this one, and remove WeightedL1. then see below

Reply:
I'd use penalty for the non convex one, then patch before and after fitting. Don't know if it's super clean... But otherwise can't you use solver.solver rather than super().fit ?